Alt text

stevenaproctor opened the pull request:
Added alt text guidance based on best practices found in [WebAim](https://webaim.org/techniques/alttext/), [AxessLab](https://axesslab.com/alt-texts/), [Deque](https://www.deque.com/blog/great-alt-text-introduction/), [Moz](https://moz.com/learn/seo/alt-text) and [Invotra](https://www.invotra.com/blogs/simple-recommendations-writing-text-descriptions-make-big-difference).

Adding a couple of extra examples may be a good idea too.

nickcolley commented on src/styles/images/index.md.njk:
```diff
@@ -25,10 +25,32 @@ If your image represents something physical, such as a letter, document or credi
 {{ example({group: "styles", item: "images", example: "alt-text", html: true, open: true, size: "l"}) }}
 
-If your image is purely decorative and a description would be confusing you should still include the alt text, but leave it empty so it will be ignored by screen readers.
+So an image will be ignored by screen readers, use `alt=""` if it is:
```
> screen readers

We could consider using the more broad term 'assistive technology' but granted we haven't really explained this term anywhere else.

(There has been some talk about having an accessibility page, which could explain this)

stevenaproctor replied:
@nickcolley I would stick with screen reader because it is not used by any other assistive technology. Looking at that sentence again, I would change it anyway.

Some images do not need descriptions. Use `alt=""` if the image is:
- decorative and does not include important content
- an icon that has a text label
- part of a link

What do you think?

nickcolley replied:
Nice 👍

stevenaproctor commented on src/styles/images/index.md.njk:
```diff
@@ -25,10 +25,32 @@ If your image represents something physical, such as a letter, document or credi
 ## Alt text
 
-All images must have an alt text. You should include a description of the image inside the tag to help users who are unable to see it.
+All images must have alt text. This is a description of an image that is read out by screen readers or is displayed if an image does not load or images have been switched off. The description tells people who cannot see an image what information the image provides.
```
@nickcolley That is a good point. I will change to 'The description tells people what information the image provides.' What do you think?

nickcolley replied:
I think without a full explanation separately to users of assistive technologies to link to being too generic might be worse than being overly specific?

What you've said is correct and better than it was.

stevenaproctor commented on src/styles/images/index.md.njk:
```diff
@@ -25,10 +25,32 @@ If your image represents something physical, such as a letter, document or credi
+
+Avoid images that contain text, especially PNGs and JPGs because the text becomes unreadable when the user zooms in. If the image does contain text, use SVG because it remains readable when the user zooms in. The alt should be the same as the text displayed in the image.
+
+If the image needs a detailed description, for example a graph or pie chart, include it in the content of the screen because it will help everyone. Use the `aria-describedby` attribute to connect the image and the description for screen reader users.
```
@nickcolley I understand your caution.

I got this from https://www.invotra.com/blogs/simple-recommendations-writing-text-descriptions-make-big-difference

We could include an example and say not to do this for the entire page's content.

nickcolley commented on src/styles/images/alt-text/index.njk:
```diff
@@ -6,4 +6,4 @@ stylesheets:
 ---
 
-<img src="larry-the-cat.jpg" alt="An image of Larry the cat, Chief Mouser to the Cabinet Office">
+<img src="larry-the-cat.jpg" alt="Larry the cat, Chief Mouser to the Cabinet Office, sitting on a meeting table wearing a Union Jack bowtie.">
```
🐱 :bowtie:

stevenaproctor replied:
😆

selfthinker commented on src/styles/images/index.md.njk:
```diff
@@ -25,10 +25,30 @@ If your image represents something physical, such as a letter, document or credi
 ## Alt text
 
-All images must have an alt text. You should include a description of the image inside the tag to help users who are unable to see it.
+All images must have alt text. This is a description of an image that tells people what information the image provides. It is read out by screen readers or displayed if an image does not load or if images have been switched off.
```
I would change `description of an image` to `meaning of an image`. The same image used in two different contexts can mean something different. For example, an icon of an eye usually means "Show" or "Preview" but a pure description would say "eye" which is less useful.

selfthinker replied:
I think that's better. I guess it's also supported by the below "describe the content and function of the image".
Although this is a point misunderstood by a lot of people, so might be worth making as clear as possible. I like how the GOV.UK guidance says somewhere (cannot find the link right now) to pretend to be on the phone and read the page aloud and how you would describe an image then.

selfthinker commented on src/styles/images/index.md.njk:
```diff
@@ -25,10 +25,30 @@ If your image represents something physical, such as a letter, document or credi
+
+- describe the content and function of the image
+- be specific, meaningful and concise
+- use normal punctuation and end with a full stop
```
Why end with a full stop? I've never heard that before.

selfthinker replied:
Do any of the articles mention why to do that, though?
I've just found this [tip from the W3C](https://www.w3.org/WAI/tutorials/images/tips/):

> As for any text, using punctuation in the text alternative makes the information easier to understand.

I guess it often makes sense. Although I don't think it is a "must". For example, alt texts for icons would be weird as a sentence.

selfthinker commented on src/styles/images/index.md.njk:
```diff
@@ -25,10 +25,30 @@ If your image represents something physical, such as a letter, document or credi
+Some images do not need descriptions. Use `alt=""` if the image is:
+
+- decorative and does not include important content
+- an icon that has a text label
```
I first misunderstood this to mean an image with a text label on it. I'm sure that's not meant. Maybe clarifying and/or adding an example would be good.

selfthinker commented on src/styles/images/index.md.njk:
```diff
@@ -25,10 +25,30 @@ If your image represents something physical, such as a letter, document or credi
+
+- decorative and does not include important content
+- an icon that has a text label
+- part of a link
```
I also misunderstood this one at first reading. I assume what is meant here is if a link consists of text and also an image.
It might be good to add to the Do's above that when an image is inside a link it has to have a description, otherwise the image will be empty to screen readers.

selfthinker commented:
I have left a few comments. I very nearly opened a PR about our alt text when I searched and found part of what I wanted to change is already here. I noticed that all of our alt text on the Images page are not great:

* The alt text example says "image of" (which this PR addresses)
* The aspect ratio example gives an unnecessary description, it nearly qualifies for an empty alt text but because there is text on the image and it's not repeated (in the example iframe itself), that text should be used ("3:2" or maybe "3 by 2 ratio")
* The representative imagery example should just have an empty alt text as it's purely decorative

I can open another PR for the second and third example. But maybe it fits better in here, as this was already started?

@accessiblewebuk is currently working on some alt text guidance for other purposes, he might have something to add?

By the way, the best resource about alt text that I know of is [W3C's alt decision tree](https://www.w3.org/WAI/tutorials/images/decision-tree/). If you ever think it makes sense to include a link to an external resource, I would highly recommend that one.

stevenaproctor commented:
@selfthinker I had added responses to your comments. Let me know what you think.

If @accessiblewebuk has anything to add that would be great.

I can amend the alt text of the aspect ratio example and representative imagery examples. I think adding the link to the W3C decision tree makes sense too.

stevenaproctor commented:
@selfthinker I have made some changes to try and address your concerns.

accessiblewebuk commented:
I also have never heard of adding a full stop to alternative text. It's certainly not a requirement for the Web Content Accessibility guidelines. 

I think "All images must have an alt text" should be modified. The term "alternative text" would be better as alt is just the attribute name (although admittedly widely known). Although it says further down that decorative images don't need alternative text it might be better here to say "All images except decorative ones must have alternative text". The alt attribute is one way of assigning alternative text but isn't the only one. It is acceptable to treat an image as decorative if the information conveyed by it is within other text on the page. 

Where it says "part of a link" under Use alt="", that depends on whether the link contains all the information conveyed by the image.

accessiblewebuk commented:
I am also not sure the best location for this information. It is in the styles section but it is nothing to do with styles. Not sure it would fit under the other categories either though.

stevenaproctor commented:
@accessiblewebuk Thanks for the comments. I have taken the full stop reference out of 'must' and added it as a separate paragraph about using normal punctuation.

I have changed the heading and beginning of the content to say 'Alternative text'.

I looked at the W3C decision tree @selfthinker linked to and have amended the 'part of a link' point to say 'used in a link and the image is not needed to understand the link'.

I agree about this may not be the best place for the guidance. But there is already some content on the images guidance and all it says is 'All images must have an alt text. You should include a description of the image inside the tag to help users who are unable to see it.' The information in other guidance or the Service Manual is similarly brief.

A few people needed some guidance so putting something in the Design System will help get the conversation started.

Let me know what you think.

amyhupe commented:
This is looking great. Thanks @stevenaproctor for making these updates and to everyone for your comments and feedback.
